br/pkg/lightning/backend/external/util.go

```diff
@@ -42,68 +44,105 @@ func seekPropsOffsets(
 	starts []kv.Key,
 	paths []string,
 	exStorage storage.ExternalStorage,
-	checkHotSpot bool,
 ) (_ [][]uint64, err error) {
 	logger := logutil.Logger(ctx)
 	task := log.BeginTask(logger, "seek props offsets")
 	defer func() {
 		task.End(zapcore.ErrorLevel, err)
 	}()
 
-	// adapt the NewMergePropIter argument types
-	multiFileStat := MultipleFilesStat{Filenames: make([][2]string, 0, len(paths))}
-	for _, path := range paths {
-		multiFileStat.Filenames = append(multiFileStat.Filenames, [2]string{"", path})
+	offsetsPerFile := make([][]uint64, len(paths))
+	for i := range offsetsPerFile {
+		offsetsPerFile[i] = make([]uint64, len(starts))
 	}
-	iter, err := NewMergePropIter(ctx, []MultipleFilesStat{multiFileStat}, exStorage, checkHotSpot)
-	if err != nil {
+	// if **all** files' first key is smaller than start key, we may lost the data
+	// between [start key, min(first key of all files)). So record this case for each
```

Comment:
larger? if all `first key in file` < `start key`, this range is invalid

Reply:
why data lost, shouldn't we allow some file doesn't have kv in some range?

Reply:
https://github.com/pingcap/tidb/pull/51676/commits/82a98c95f14984fdc9814e982ad038b5fb37526c I have added some comments, please check if it's clear enough.

We can allow part of files does not cover the first `starts` key, but not all files.